File: app/src/main/java/org/cloud/sonic/android/plugin/touchPlugin/touch/WindowManagerWrapper.java
package org.cloud.sonic.android.plugin.touchPlugin.touch;

import android.os.RemoteException;
import android.view.IRotationWatcher;

import org.cloud.sonic.android.utils.InternalApi;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * More https://github.com/DeviceFarmer/STFService.apk
 */
public class WindowManagerWrapper {
    private Object windowManager;

    public static interface RotationWatcher {
        public void onRotationChanged(int rotation);
    }

    public WindowManagerWrapper() {
        windowManager = getWindowManager();
    }

    public int getRotation() {
        try {
            Method getter = windowManager.getClass().getMethod("getDefaultDisplayRotation");
            return (Integer) getter.invoke(windowManager);
        }
        catch (NoSuchMethodException e) {}
        catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        catch (InvocationTargetException e) {
            e.printStackTrace();
        }

        try {
            Method getter = windowManager.getClass().getMethod("getRotation");
            return (Integer) getter.invoke(windowManager);
        }
        catch (NoSuchMethodException e) {
            e.printStackTrace();
        }
        catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        catch (InvocationTargetException e) {
            e.printStackTrace();
        }

        return 0;
    }

    public static Object getWindowManager() {
        return InternalApi.getServiceAsInterface("window", "android.view.IWindowManager$Stub");
    }

    public Object watchRotation(final RotationWatcher watcher) {
        IRotationWatcher realWatcher = new IRotationWatcher.Stub() {
            @Override
            public void onRotationChanged(int rotation) throws RemoteException {
                watcher.onRotationChanged(rotation);
            }
        };

        try {
            Method getter = windowManager.getClass().getMethod("watchRotation", IRotationWatcher.class, int.class);
            getter.invoke(windowManager, realWatcher, 0);
            return realWatcher;
        }
        catch (NoSuchMethodException e) {
            try {
                Method getter = windowManager.getClass().getMethod("watchRotation", IRotationWatcher.class);
                getter.invoke(windowManager, realWatcher);
                return realWatcher;
            }
            catch (NoSuchMethodException e2) {
                throw new UnsupportedOperationException("watchRotation is not supported: " + e2.getMessage());
            }
            catch (IllegalAccessException e2) {
                throw new UnsupportedOperationException("watchRotation is not supported: " + e2.getMessage());
            }
            catch (InvocationTargetException e2) {
                throw new UnsupportedOperationException("watchRotation is not supported: " + e2.getMessage());
            }
        }
        catch (IllegalAccessException e) {
            throw new UnsupportedOperationException("watchRotation is not supported: " + e.getMessage());
        }
        catch (InvocationTargetException e) {
            throw new UnsupportedOperationException("watchRotation is not supported: " + e.getMessage());
        }
    }


}

File: app/src/main/java/org/cloud/sonic/android/plugin/touchPlugin/touchCompat/WindowManagerWrapper.kt
/*
 *  sonic-android-apk  Help your Android device to do more.
 *  Copyright (C) 2022 SonicCloudOrg
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.cloud.sonic.android.plugin.touchPlugin.touchCompat

import android.os.IBinder
import android.os.RemoteException
import android.view.IRotationWatcher
import java.lang.reflect.InvocationTargetException

/**
 * @see https://github.com/openstf/STFService.apk/blob/master/app/src/main/java/jp/co/cyberagent/stf/compat/WindowManagerWrapper.java
 */
class WindowManagerWrapper {
    private var windowManager: Any? = null

    init {
        windowManager = getWindowManager()
    }

    companion object {
        private fun getBinder(name: String): Any {
            val serviceManager = Class.forName("android.os.ServiceManager")
            val getService = serviceManager.getMethod(
                "getService",
                String::class.java
            )
            return getService.invoke(null, name)
        }

        private fun getService(serviceName: String, interfaceClass: String): Any {
            val serviceBinder = getBinder(serviceName)
            val s = Class.forName(interfaceClass)
            val asInterface = s.getMethod("asInterface", IBinder::class.java)
            return asInterface.invoke(null, serviceBinder)
        }

        fun getWindowManager(): Any? {
            return getService("window", "android.view.IWindowManager\$Stub")
        }

    }

    fun watchRotation(watcher: RotationWatcher): Any {
        val realWatcher: IRotationWatcher = object : IRotationWatcher.Stub() {
            @Throws(RemoteException::class)
            override fun onRotationChanged(rotation: Int) {
                watcher.onRotationChanged(rotation)
            }
        }

        return try {
            val getter = windowManager!!.javaClass.getMethod(
                "watchRotation",
                IRotationWatcher::class.java,
                Int::class.javaPrimitiveType
            )
            getter.invoke(windowManager, realWatcher, 0)
            realWatcher
        } catch (e: NoSuchMethodException) {
            try {
                val getter = windowManager!!.javaClass.getMethod(
                    "watchRotation",
                    IRotationWatcher::class.java
                )
                getter.invoke(windowManager, realWatcher)
                realWatcher
            } catch (e2: java.lang.Exception) {
                throw UnsupportedOperationException("UnsupportedOperationException: " + e2.message)
            }
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        }
    }

    fun getRotation(): Int {
        try {
            val getter = windowManager!!.javaClass.getMethod("getDefaultDisplayRotation")
            return getter.invoke(windowManager) as Int
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        try {
            val getter = windowManager!!.javaClass.getMethod("getRotation")
            return getter.invoke(windowManager) as Int
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return 0
    }

    interface RotationWatcher {
        fun onRotationChanged(rotation: Int)
    }
}
